lib/IDE/ArgumentCompletion.cpp

```diff
@@ -176,12 +237,40 @@ void ArgumentTypeCheckCompletionCallback::sawSolutionImpl(const Solution &S) {
   llvm::SmallDenseMap<const VarDecl *, Type> SolutionSpecificVarTypes;
   getSolutionSpecificVarTypes(S, SolutionSpecificVarTypes);
 
-  Results.push_back({ExpectedTy, isa<SubscriptExpr>(ParentCall), Info.Value,
-                     Info.ValueTy, ArgIdx, ParamIdx, std::move(ClaimedParams),
+  Results.push_back({ExpectedTy, ExpectedCallType,
+                     isa<SubscriptExpr>(ParentCall), Info.Value, Info.ValueTy,
+                     ArgIdx, ParamIdx, std::move(ClaimedParams),
                      IsNoninitialVariadic, Info.BaseTy, HasLabel, IsAsync,
                      SolutionSpecificVarTypes});
 }
 
+void ArgumentTypeCheckCompletionCallback::computeOverriddenDecls(
+    SmallPtrSetImpl<ValueDecl *> &OverriddenDecls) {
+  for (size_t i = 0; i < Results.size(); ++i) {
+    auto &ResultA = Results[i];
+    for (size_t j = i + 1; j < Results.size(); ++j) {
+      auto &ResultB = Results[j];
+      if (!ResultA.FuncD || !ResultB.FuncD) {
+        continue;
+      }
+      if (ResultA.FuncD->getName() != ResultB.FuncD->getName()) {
+        continue;
+      }
+      bool aAsSpecializedAsB =
+          isDeclAsSpecializedAs(DC, ResultA.FuncD, ResultB.FuncD);
+      bool bAsSpecializedAsA =
+          isDeclAsSpecializedAs(DC, ResultB.FuncD, ResultA.FuncD);
```

Comment:
We cannot really determine whether something is better without considering the context (i.e. arguments to the call where these overloads are used), the logic we have in the solver at the moment uses it only when there are exactly two overload choices and even then there are cases where it's wrong.

Reply:
The motivating example was the following 

```swift
protocol MyProtocol {
  init(init2: Int)
}

extension MyProtocol {
  init(init2: Int) { ... }
}

class MyClass: MyProtocol {
  init(init2: Int) { ... }
}

func testConstructer() {
  MyClass(#^COMPLETE^#)
}
```

where we don’t want to suggest the `init(init2:)` from the protocol extension. 

So what I’ve done now is to just extract the logic that checks if we have two functions with the same name and same type where one is defined in a protocol extension and the other is not.